### PR TITLE
Add CartesianRepresentationFrameAttribute to GCRS and PrecessedGeocentric frames

### LINKED ISSUE #4926: Cannot create GCRS frames with obsgeoloc shapes other than (3,)
After discussion with @eteq, the code that was merged for #4890 included the use of GCRS frames whose `obsgeoloc` and `obsgeovel` vectors could have shapes equal to `(3, N)`, where `N` is the number of times at which the routine `EarthLocation.get_gcrs_posvel` calculates the position and velocity of an Earth-based observer.

However, in https://github.com/astropy/astropy/blob/3c47c1067036a6cf3dc13ebc0fc268948d4ccc29/astropy/coordinates/builtin_frames/gcrs.py#L74, the shape of the obsgeoloc arrays are hardcoded to `(3,)`.

This means that any attempt to call `coordinates.get_body` with a non-scalar time and a defined location will raise an error

``` python
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-12-1102973f833a> in <module>()
----> 1 coord.get_body(tarr2, 'sun', location=wht)

/usr/local/lib/python3.5/site-packages/astropy/coordinates/solar_system.py in get_body(time, body, location)
    252         gcrs = icrs.transform_to(GCRS(obstime=time,
    253                                       obsgeoloc=obsgeoloc,
--> 254                                       obsgeovel=obsgeovel))
    255     else:
    256         gcrs = icrs.transform_to(GCRS(obstime=time))

/usr/local/lib/python3.5/site-packages/astropy/coordinates/baseframe.py in __init__(self, *args, **kwargs)
    482 
    483             # Validate input by getting the attribute here.
--> 484             getattr(self, fnm)
    485 
    486         args = list(args)  # need to be able to pop them

/usr/local/lib/python3.5/site-packages/astropy/coordinates/baseframe.py in __get__(self, instance, frame_cls)
    209             out = self.default
    210 
--> 211         out, converted = self.convert_input(out)
    212         if instance is not None and converted:
    213             setattr(instance, '_' + self.name, out)

/usr/local/lib/python3.5/site-packages/astropy/coordinates/baseframe.py in convert_input(self, value)
    332                 raise ValueError('The provided value has shape "{0}", but '
    333                                  'should have shape "{1}"'.format(value.shape,
--> 334                                                                   self.shape))
    335             if (oldvalue.unit == value.unit and hasattr(oldvalue, 'value') and
    336                 np.all(oldvalue.value == value.value)):

ValueError: The provided value has shape "(3, 2)", but should have shape "(3,)"
```

### PR DESCRIPTION
**Pull request by StuartLittlefair**
This PR fixes #4926, and is a replacement for #5196.

As @mhvk mentioned on the latter PR, it would be much more logical for the ```obsgeoloc``` and ```obsgeovel``` attributes to be ```CartesianRepresentations```. 

I was initially reluctant, but I realised that you can be flexible about what types of values initialise these attributes so that (3, ...) or (..., 3) shaped ```Quantities``` are allowed, as well as ```CartesianRepresentations``` themselves. This means the code changes are minimal, and anyone who sets these attributes in their code will find their code still works.

### CODE REVIEW

**Comment by StuartLittlefair** [comment]
I should mention it would be nice to merge this ASAP, since astropy/astroplan#167 introduced some fairly dodgy hacks to workaround #4926. It would be nice to remove those before the next version of astroplan is released.

**Comment by mhvk** [comment]
@StuartLittlefair - This looks good, except that I think we should *not* aim to be flexible in where the three coordinates are located: I think they have to be the first dimension, as for `CartesianRepresentation` itself. Otherwise what does a `(3, 3)` input mean?

**Comment by mhvk** [comment]
p.s. Agreed it would be nice to merge this soon -- it will also help make `baseframe` shape-aware.

**Comment by mhvk** [comment]
Also, this does need a changelog entry.  Is this something that makes sense to backport? My guess is not, so I put the milestone at 1.2.2 for now.

**Comment by mhvk** [comment]
p.p.p.s. The failures seem real.

**Comment by StuartLittlefair** [comment]
Ok @mhvk - I can see that (3,3) is ambiguous. It's a one liner to change this and most code within astropy shouldn't need to change. There may be some potential changes to ```get_gcrs_posvel``` which is publically exposed. It might be better to leave that as is, and reshape the returned quantities before creating ```GCRS``` frames from them. This is a little clunky though!

I'm not sure about backporting. It would be useful for astroplan to backport to 1.2 I think, but I don't fully understand the implications. 

I'll have to look at that and the failures tomorrow morning.

**Comment by StuartLittlefair** [comment]
@mhvk good call on removing the flexibility. Everything "just works" regardless. 

The only user-facing change as a result of this PR is that accessing the ```obsgeoloc``` attribute now returns a ```CartesianRepresentation``` so someone may have to replace ```coord.obsgeoloc``` with ```coord.obsgeoloc.xyz``` in their code, but that's the only outward change I can see.

File: astropy/coordinates/baseframe.py [hunk 1]
```diff
@@ -276,6 +276,66 @@ def convert_input(self, value):
...
+
+        Raises
+        ------
+        ValueError
+            If the input is not valid for this attribute.
+        """
+        if np.all(value == 0) and self.unit is not None:
+            return CartesianRepresentation(np.zeros(3) * self.unit), True
+        else:
+            converted = True
```
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: At this point, can we just do
```
cartrep = CartesianRepresentation(value)
```
I.e., do we need a unit change at all? It seems a lot of code for something that I find hard to see the use of...

**Comment by mhvk** [comment]
@StuartLittlefair - great that it "just works"!! I have one in-line question for simplification. Another is a more general one: why would `obsgeoloc` at least not use `EarthLocation`? (for which there is already an attribute).

In the end, I don't think it matters too much (indeed, `EarthLocation` should probably start using `CartesianRepresentation` inside...), and probably I should just read the code better to understand what these attributes are for... 

In any case, it would probably be good if some of those who wrote the code can have a quick look as well, so cc @taldcroft, @eteq.

File: astropy/coordinates/baseframe.py (hunk 1 above)
Comment (StuartLittlefair) [inline astropy/coordinates/baseframe.py]: The idea is to stop someone setting the attributes with a ```Quantity``` or ```CartesianRepresentation``` that's in the wrong units. 

The other purpose is that later in the coordinate transformation code we use the fact that we can trust ```obsgeoloc``` to be in metres and ```obsgeovel``` to be in metres/sec.

**Comment by StuartLittlefair** [comment]
Thanks for the suggestion @mhvk . I was quite impressed by how simple it was. Almost as if someone thought carefully when they wrote these objects in the first place ;)

One reason not to use ```EarthLocation``` is that ```obsgeoloc``` is a coordinate in the ```GCRS``` frame, whereas ```EarthLocation``` is in a different frame (```ITRS```?). If we had an ```EarthLocationFrameAttribute``` it would require a transformation every time we needed the ```GCRS``` coordinate.

File: astropy/coordinates/baseframe.py (hunk 1 above)
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: OK, it sounds like `CartesianRepresentation` should obtain a `.to` method for unit conversion... But that can be in a separate PR...

**Comment by mhvk** [comment]
Also OK on `EarthLocation` -- at some point I just have to bite the bullet and try to understand the differences between the difference frames...

File: astropy/coordinates/baseframe.py [hunk 2]
```diff
@@ -276,6 +276,66 @@ def convert_input(self, value):
...
+            # if it's a CartesianRepresentation, get the xyz Quantity
+            value = getattr(value, 'xyz', value)
+            if not (hasattr(value, 'unit') ):
+                raise TypeError('Tried to set a CartesianRepresentationFrameAttribute with '
+                                'something that does not have a unit.')
+            oldvalue = value
+            value = u.Quantity(oldvalue, copy=False).to(self.unit)
```
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: Couldn't this just be?
```
value = u.Quantity(oldvalue, self.unit, copy=False)
```

Comment (StuartLittlefair) [inline astropy/coordinates/baseframe.py]: Yes!

**Comment by StuartLittlefair** [comment]
This PR fixes #5209

**Comment by StuartLittlefair** [comment]
This PR almost fixes #5209. Instead the example there causes a different bug:

```python
In [4]: SkyCoord(moon)
Out[4]: 
<SkyCoord (GCRS: obstime=2014-01-20 22:00:00.000, obsgeoloc=(1712675.0, 5328049.0, -3053885.0) m, obsgeovel=(-388.0, 125.0, 0.0) m / s): (ra, dec, distance) in (deg, deg, AU)
    (172.4, 0.51, 0.002671)>

In [5]: SkyCoord([moon])
Out[5]: 
<SkyCoord (ICRS): (ra, dec) in deg
    (172.4, 0.51)>

In [6]: SkyCoord([moon, moon])
Out[6]: 
<SkyCoord (ICRS): (ra, dec, distance) in (deg, deg, AU)
    [(172.4, 0.51, 0.002671), (172.4, 0.51, 0.002671)]>

In [7]: SkyCoord([moon.icrs, moon.icrs])
Out[7]: 
<SkyCoord (ICRS): (ra, dec, distance) in (deg, deg, AU)
    [(122.91386538, 19.98343108, 0.98321877),
     (122.91386538, 19.98343108, 0.98321877)]>

In [8]: moon.icrs
Out[8]: 
<SkyCoord (ICRS): (ra, dec, distance) in (deg, deg, AU)
    (122.91386538, 19.98343108, 0.98321877)>
```

File: astropy/coordinates/baseframe.py [hunk 3]
```diff
@@ -276,6 +276,66 @@ def convert_input(self, value):
...
+                raise TypeError('Tried to set a CartesianRepresentationFrameAttribute with '
+                                'something that does not have a unit.')
+            oldvalue = value
+            value = u.Quantity(oldvalue, self.unit, copy=False)
+            if (oldvalue.unit == value.unit and hasattr(oldvalue, 'value') and
+                np.all(oldvalue.value == value.value)):
+                converted = False
+
+            # now try and make a CartesianRepresentation. Aim to be flexible
+            cartrep = CartesianRepresentation(value)
```
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: Here, I think you can add `copy=False`, no?

File: astropy/coordinates/baseframe.py [hunk 4]
```diff
@@ -276,6 +276,66 @@ def convert_input(self, value):
...
+            # if it's a CartesianRepresentation, get the xyz Quantity
+            value = getattr(value, 'xyz', value)
+            if not (hasattr(value, 'unit') ):
+                raise TypeError('Tried to set a CartesianRepresentationFrameAttribute with '
+                                'something that does not have a unit.')
+            oldvalue = value
+            value = u.Quantity(oldvalue, self.unit, copy=False)
+            if (oldvalue.unit == value.unit and hasattr(oldvalue, 'value') and
+                np.all(oldvalue.value == value.value)):
+                converted = False
```
Comment (mhvk) [inline astropy/coordinates/baseframe.py]: As I understand `FrameAttribute`, converted should be true if any work has been done, so that the "reworked" value can be stored and retrieved later without again converting. Since here you always create a new `CartesianRepresentation`, this suggests `convert` should always be `True`. (@taldcroft: since you wrote this code: is this correct?)

If the above holds, it might be an idea to check whether a given value already is a `CartesianRepresentation` with the correct unit, since that would be the only case for which no conversion would be necessary.

File: astropy/coordinates/builtin_frames/gcrs.py [hunk 5]
```diff
@@ -32,14 +33,14 @@ class GCRS(BaseCoordinateFrame):
         The time at which the observation is taken.  Used for determining the
         position of the Earth.
     * ``obsgeoloc``
-        Vector giving the position of the observer relative to the
-        center-of-mass of the Earth, oriented the same as BCRS/ICRS.
-        The leading dimension of this vector needs to be 3.
+        The position of the observer relative to the center-of-mass of the Earth,
+        oriented the same as BCRS/ICRS.
+        Either 0, a CartesianRepresentation, or a Quantity of shape (3, ...)
```
Comment (mhvk) [inline astropy/coordinates/builtin_frames/gcrs.py]: Not super-important, but we've tended to give fully linkable entries to astropy classes in the docstrings, i.e., something like
```
        The position of the observer relative to the center-of-mass of the Earth,
        oriented the same as BCRS/ICRS. Either 0, `~astropy.coordinates.CartesianRepresentation`,
        or proper input for one, i.e., a `~astropy.units.Quantity` with shape (3, ...) and length units.
```

File: astropy/coordinates/builtin_frames/icrs_cirs_transforms.py [hunk 6]
```diff
@@ -119,8 +119,10 @@ def cirs_to_cirs(from_coo, to_frame):
 @frame_transform_graph.transform(FunctionTransform, ICRS, GCRS)
 def icrs_to_gcrs(icrs_coo, gcrs_frame):
     # first set up the astrometry context for ICRS<->GCRS
-    pv = np.array([gcrs_frame.obsgeoloc.value,
-                   gcrs_frame.obsgeovel.value])
+    pv = np.array([gcrs_frame.obsgeoloc.xyz.value,
+                   gcrs_frame.obsgeovel.xyz.value])
+    if pv.ndim > 2:
+        pv = np.rollaxis(pv, -1, 0)
```
Comment (mhvk) [inline astropy/coordinates/builtin_frames/icrs_cirs_transforms.py]: I think this will not work for multidimensional locations/velocities: you want axis `0` and `1` both to go to the end, no? I.e., `np.rollaxis(np.rollaxis(pv, 0, pv.ndim), 0, pv.ndim)`

p.s. I was thinking it might be cleaner to just do it in the definition of `pv`, but that is also rather messy
```
loc = gcrs_frame.obsgeoloc.xyz.value
vel = gcrs_frame.obsgeovel.xyz.value
pv = np.stack((np.rollaxis(loc, 0, loc.ndim), np.rollaxis(vel, 0, vel.ndim)), axis=-2)
```

p.s.2 Maybe we should backport `np.moveaxes`...

File: astropy/coordinates/representation.py [hunk 7]
```diff
@@ -223,6 +223,24 @@ def _unitstr(self):
                            for component in self.components]))
         return unitstr
 
+    def __eq__(self, other):
```
Comment (mhvk) [inline astropy/coordinates/representation.py]: Just to be sure: do we really need this? I ask in part because I think a representation is more like an array, and if I compared two instances, I would thus expect to get an array of bool back.

**Comment by mhvk** [comment]
@StuartLittlefair - sorry for commenting in bits and pieces, but there are more now. Perhaps the most serious one is that I think comparison between representations should give back an array of bool. Where is this needed?

File: astropy/coordinates/representation.py (hunk 7 above)
Comment (StuartLittlefair) [inline astropy/coordinates/representation.py]: This is needed because of some checks the coordinate machinery does initialising frames - two frames with identical ```obsgeoloc``` attributes need to satisfy ```np.any(obsgeoloc1 == obsgeoloc2)```

However, that would also  be alright if this returned an array...

File: astropy/coordinates/baseframe.py (hunk 4 above)
Comment (StuartLittlefair) [inline astropy/coordinates/baseframe.py]: I tried that, but I got lots of ```CartesianRepresentation cannot be indexed``` errors that I didn't understand, so backed away...

**Comment by StuartLittlefair** [comment]
Looks like the failures are related to np.stack not being available in numpy <1.10. 

I'll go back to a rollaxis solution but this won't get done today now. Sorry!

**Comment by StuartLittlefair** [comment]
@mhvk it looks like the one failure was just a timeout - could you restart?

File: astropy/coordinates/baseframe.py (hunk 4 above)
Comment (taldcroft) [inline astropy/coordinates/baseframe.py]: @mhvk - your comment about the way `converted` works is correct.  If `convert_input` returns anything besides the original `value` object then `converted` should be `True`.

File: astropy/coordinates/baseframe.py [hunk 8]
```diff
@@ -276,6 +276,66 @@ def convert_input(self, value):
...
+    default : object
+        Default value for the attribute if not provided
+    secondary_attribute : str
+        Name of a secondary instance attribute which supplies the value if
+        ``default is None`` and no value was supplied during initialization.
+    unit : unit object or None
+        Name of a unit that the input will be converted into. If None, no
+        unit-checking or conversion is performed
+    """
+    def __init__(self, default=None, secondary_attribute='', unit=None, shape=None):
```
Comment (taldcroft) [inline astropy/coordinates/baseframe.py]: `shape` seems orphaned here.

File: astropy/coordinates/baseframe.py [hunk 9]
```diff
@@ -276,6 +276,66 @@ def convert_input(self, value):
...
+        out, converted : correctly-typed object, boolean
+            Tuple consisting of the correctly-typed object and a boolean which
+            indicates if conversion was actually performed.
+
+        Raises
+        ------
+        ValueError
+            If the input is not valid for this attribute.
+        """
+        if np.all(value == 0) and self.unit is not None:
```
Comment (taldcroft) [inline astropy/coordinates/baseframe.py]: Just wondering about this "special value".  It seems like it makes code slightly more difficult to read.  If I saw `obsgeoloc = 0` then I would be confused because `0` is not a valid cartesian point.  If the `0` just to stop from having to type `[0, 0, 0] * u.m`, then perhaps explicit is better here?

File: astropy/coordinates/baseframe.py [hunk 10]
```diff
@@ -276,6 +276,66 @@ def convert_input(self, value):
...
+            return CartesianRepresentation(np.zeros(3) * self.unit), True
+        else:
+            converted = True
+            # if it's a CartesianRepresentation, get the xyz Quantity
+            value = getattr(value, 'xyz', value)
+            if not (hasattr(value, 'unit') ):
+                raise TypeError('Tried to set a CartesianRepresentationFrameAttribute with '
```
Comment (taldcroft) [inline astropy/coordinates/baseframe.py]: Preferred convention for new code is to start exception messages with lower case (follows Python core).

File: astropy/coordinates/baseframe.py [hunk 11]
```diff
@@ -276,6 +276,66 @@ def convert_input(self, value):
...
+                                'something that does not have a unit.')
+            oldvalue = value
```
Comment (taldcroft) [inline astropy/coordinates/baseframe.py]: I haven't thought through the logic super carefully, but wondering if you could you just do:
```
if value.unit != self.unit:
    value = value.to(self.unit)
    converted = True  # and remove the initialization above
else:
    convert = False
```

File: astropy/coordinates/builtin_frames/gcrs.py [hunk 12]
```diff
@@ -70,8 +71,8 @@ class GCRS(BaseCoordinateFrame):
     default_representation = SphericalRepresentation
 
     obstime = TimeFrameAttribute(default=DEFAULT_OBSTIME)
-    obsgeoloc = QuantityFrameAttribute(default=0, unit=u.m, shape=(3,))
-    obsgeovel = QuantityFrameAttribute(default=0, unit=u.m/u.s, shape=(3,))
+    obsgeoloc = CartesianRepresentationFrameAttribute(default=0, unit=u.m)
```
Comment (taldcroft) [inline astropy/coordinates/builtin_frames/gcrs.py]: Per previous comment, this could be `CartesianRepresentationFrameAttribute(default=[0, 0, 0], unit=u.m)` (I think).

**Comment by StuartLittlefair** [comment]
Thanks for the comments @taldcroft 

The last few commits should address them, I think.

**Comment by StuartLittlefair** [comment]
Fixes #5216

File: CHANGES.rst [hunk 13]
```diff
@@ -178,6 +178,9 @@ Bug Fixes
 
 - ``astropy.coordinates``
 
+  - Changed obsgeoloc and obsgeovel frame attributes to CartesianRepresentation. Fixes creation of
```
Comment (taldcroft) [inline CHANGES.rst]: I wonder if this should also be listed in API changes.

File: astropy/coordinates/baseframe.py [hunk 14]
```diff
@@ -276,6 +276,68 @@ def convert_input(self, value):
...
+        out, converted : correctly-typed object, boolean
+            Tuple consisting of the correctly-typed object and a boolean which
+            indicates if conversion was actually performed.
+
+        Raises
+        ------
+        ValueError
+            If the input is not valid for this attribute.
+        """
+        if (isinstance(value, list) and np.all([v == 0 for v in value]) and
```
Comment (taldcroft) [inline astropy/coordinates/baseframe.py]: What about `np.all(value == [0, 0, 0]) and self.unit is not None`.  That allows for a numpy 3-vector as well.

**Comment by taldcroft** [comment]
Overall this is looking close.  I just looked at the test diffs and it does not appear that all of the logical code paths are being tested.  I'm guessing that existing tests won't cover this, but I haven't explicitly looked.

File: astropy/coordinates/baseframe.py (hunk 14 above)
Comment (StuartLittlefair) [inline astropy/coordinates/baseframe.py]: I don't think that works since if ```value``` is a CartesianRepresentation it will raise an exception (they cannot be indexed). 

Let me think about it.

Comment (taldcroft) [inline astropy/coordinates/baseframe.py]: Seems to work (i.e. just gives `False`), but in any case not a big deal.
```
In [8]: np.all([0,0,0] == CartesianRepresentation([0,1,0]*u.m))
Out[8]: False
```

Comment (StuartLittlefair) [inline astropy/coordinates/baseframe.py]: Great. For some reason I thought it would try and unpack the CartesianRepresentation!  Not sure why

Comment (taldcroft) [inline astropy/coordinates/baseframe.py]: But you are right in that I vaguely recall that there are some narrow circumstances where doing a comparison like that can raise an exception.  Sadly I can't recall a specific example...

Comment (StuartLittlefair) [inline astropy/coordinates/baseframe.py]: Damn - this fails with Quantities that have ndim > 1.

**Comment by StuartLittlefair** [comment]
I've added a few extra lines to the regression test which should test all the logic I added.

File: CHANGES.rst (hunk 13 above)
Comment (eteq) [inline CHANGES.rst]: definitely an API change.

**Comment by eteq** [comment]
@StuartLittlefair - I like the idea here and I think this is basically good (it looks to me like you had no more changes pending, right?).

But I have one concern: backwards compatibility.  If I understand right, this is still backwards compatible at the *input* - that is, if you provide an `obsgeoloc` as ``sc = SkyCoord(..., obsgeoloc=[1,2,3]*u.km,...)`` it still works... but if I do ``sc.obsgeoloc`` I get something *different* now, so any code written to work with quantities no longer works.  I'm reading that right, correct? So the question is: can we think of any way to maintain the ~ standard practice of having at least one version of deprecation for API changes?  One option might be to use new names (`obsgeoloc`/`obsgeovel` are not great, anyway... although I have no immediate suggestions because I wrote them originally), have those be the *real* attributes, but have `obsgeoloc` and `obsgeovel` translate back to `Quantity` while raising deprecation warnings.

Alternatively, if we think these attributes are never actually accessed in the wild, then maybe it's not worth worrying much about it.  But it should still definitely be listed in the API changes section of the changelog.

**Comment by mhvk** [comment]
Now back from holidays and had a bit more of a look. On the backwards compatibility, possibly the direct use of the attributes is a rare enough case that we can just do it rather than deprecate?

I ask in part because at a bit higher level I wonder a bit about the logic of having a separate `loc` and `vel`: logically this is really an `EarthLocation` or `SkyCoord`  that includes proper motion and is set to a particular reference frame. I think eventually we might want to move to a name that reflects that (perhaps just `location` as for the `altaz` frame). But it might be good not to move more than once!

Above, @StuartLittlefair mentioned that one perhaps shouldn't use `EarthLocation` directly since then every time one needs to convert; but if so, we might want to store this as the combination of the two that is actually used in the calculation (i.e., erfa's `pv` array).

Anyway, it is all leaving me a bit confused. However, from the perspective of allowing proper broadcasting between coordinates and their frame attributes, it certainly is very helpful that with the current implementation, the `shape` attribute is correct (without the `3` for the x,y,z part).

**Comment by mhvk** [comment]
p.s. Separately, `EarthLocation` has a `get_gcrs_posvel` method. I think this will have to be changed to return a tuple of `CartesionRepresentation` as well. Which of course changes the API as well...

**Comment by StuartLittlefair** [comment]
@mhvk, @eteq:

Thanks for the comments. A few thoughts:

- I feel the question about whether to deprecate or not is slightly above my pay grade. However, my feeling is not to bother. I think doing ```sc.obsgeoloc``` is probably a very rare use case. Moreover, the change required in any code would be to replace ```sc.obsgeoloc``` with ```sc.obsgeoloc.xyz```, which is not major.

-  If we do deprecate, what about making attributes which shadow the main ones, so ```_obsgeoloc```, for example? These would actually be ```CartesianRepresentation```s so would maintain the shape advantages for broadcasting. We'd make ```obsgeoloc``` a straight property, whose getter raised a deprecation warning. Then when we finally remove the deprecation warning, changing the codebase is a matter of replacing ```_obsgeoloc``` with ```obsgeoloc```.

- On the issue of moving away from separate locations and velocities I feel this is something to revisit when the challenge of implementing proper motion is addressed. Until then, the way it is done now gives you the flexibility to make ```GCRS``` frames for orbiting satellites etc if you want to do that.

- Finally, addressing ```get_gcrs_posvel```. I thought about changing this for consistency but was in two minds. In the end I decided not to change it, since the output from ```get_gcrs_posvel``` can be supplied as is to initialise a ```GCRS``` frame and it minimised the API changes. Should I revisit that decision?

Will add an API changeling entry, but pending a decision on deprecation, that is the only outstanding change.

File: CHANGES.rst [hunk 15]
```diff
@@ -178,8 +223,9 @@ Bug Fixes
 
 - ``astropy.coordinates``
 
-  - Changed obsgeoloc and obsgeovel frame attributes to CartesianRepresentation. Fixes creation of
-    GCRS frames representing a location on Earth with multiple obstimes. [#5207]
+  - GCRS frames representing a location on Earth with multiple obstimes are now allowed. This means that
+   the solar system routines ``get_body``, ``get_moon`` and ``get_sun`` now work with non-scalar times and
```
Comment (mhvk) [inline CHANGES.rst]: I think the travis failure is because the indentation here is not exactly right (add one more space to align this line and the one below).

**Comment by StuartLittlefair** [comment]
Thanks @mhvk. This is now passing.

Pending some consensus on deprecation I am finished working on the PR.

**Comment by mhvk** [comment]
To have something to look at, I changed `EarthLocation.get_gcrs_posvel` to return `CartesianRepresentation` as well. Somehow, I couldn't make a PR to this branch (no @StuartLittlefair branch to send the request to, very odd), so I made a new PR (where I also squashed all the commits here into one, and rebased). Please have a look at #5253.

**Comment by mhvk** [comment]
Am closing this since it is part of #5253, and there seems to be agreement that it is better to change `EarthLocation.get_gcrs_posvel` as well.

File: astropy/coordinates/representation.py (hunk 7 above)
Comment (mhvk) [inline astropy/coordinates/representation.py]: @StuartLittlefair - in @eteq's review of this in #5253, the addition of `__eq__` came up again, as it breaks backward compatibility. I just checked and everything runs fine if I simply remove this again, which leaves me even more confused. My sense right now would be to leave it out. If we do want equality checks, we can do that in a separate PR.

Comment (StuartLittlefair) [inline astropy/coordinates/representation.py]: Sorry - I'm travelling on holiday so this is a bit rushed with my last bit
of internet.

I vaguely recall that there is an issue with creating GCRS frames with
obsgeoloc and obsgeovel set, and particularly set to >ND. This may have
gone away with adoption of cartesianrepresentations? If so then I'm happy
to remove _eq_

Stuart Littlefair
Dept. of Physics & Astronomy
Univ. of Sheffield, Sheffield, S3 7RH

email: s.littlefair@shef.ac.uk
Phone: +44 114 2224525
Sent from my iPhone

On 30 Aug 2016, at 21:15, Marten van Kerkwijk <notifications@github.com>
wrote:

In astropy/coordinates/representation.py
<https://github.com/astropy/astropy/pull/5207#discussion_r76870586>:

> @@ -223,6 +223,24 @@ def _unitstr(self):
>                             for component in self.components]))
>          return unitstr
>
> +    def __eq__(self, other):

@StuartLittlefair <https://github.com/StuartLittlefair> - in @eteq
<https://github.com/eteq>'s review of this in #5253
<https://github.com/astropy/astropy/pull/5253>, the addition of __eq__ came
up again, as it breaks backward compatibility. I just checked and
everything runs fine if I simply remove this again, which leaves me even
more confused. My sense right now would be to leave it out. If we do want
equality checks, we can do that in a separate PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/astropy/astropy/pull/5207/files/92f42af1864227c122b61286585e5ffcedfb36f1#r76870586>,
or mute the thread
